File: Modules/ServerTypes/TaskFlow/WorkspaceMemberDetailForm.ts
import { initFormType, PrefixedContext } from "@serenity-is/corelib";
import { WorkspaceMemberRoleEditor } from "../../TaskFlow/WorkspaceMember/WorkspaceMemberRoleEditor";
import { WorkspaceMemberUserEditor } from "../../TaskFlow/WorkspaceMember/WorkspaceMemberUserEditor";

export interface WorkspaceMemberDetailForm {
    UserId: WorkspaceMemberUserEditor;
    Role: WorkspaceMemberRoleEditor;
}

export class WorkspaceMemberDetailForm extends PrefixedContext {
    static readonly formKey = 'TaskFlow.WorkspaceMemberDetail';
    declare private static init: boolean;

    constructor(...args: ConstructorParameters<typeof PrefixedContext>) {
        super(...args);

        if (!WorkspaceMemberDetailForm.init) {
            WorkspaceMemberDetailForm.init = true;

            initFormType(WorkspaceMemberDetailForm, [
                'UserId', WorkspaceMemberUserEditor,
                'Role', WorkspaceMemberRoleEditor
            ]);
        }
    }
}

File: Modules/TaskFlow/WorkspaceMember/WorkspaceMemberDetailForm.cs
namespace TaskFlowSerenity.TaskFlow.Forms;

[FormScript("TaskFlow.WorkspaceMemberDetail")]
[BasedOnRow(typeof(WorkspaceMemberRow), CheckNames = true)]
public class WorkspaceMemberDetailForm
{
    [EditorType("TaskFlowSerenity.TaskFlow.WorkspaceMemberUserEditor")]
    public int UserId { get; set; }
    public string Role { get; set; }
}